### PR TITLE
Alternate mailing list signup process

### PR DESCRIPTION
**Pull request by peteryates**
## Trello

https://trello.com/c/3HESZ1z6/1524-redesign-the-email-sign-up-form-register-your-interest-on-one-page

## Context and changes

The current mailing list signup wizard might be too long. The plan is to test a shorter version.

This PR implements a one page signup wizard. It duplicates much of the logic from the wizard because one will be deleted once the testing's over, so there's no need to share code between them.

* [x] Create a model, controller and basic signup form
* [x] Hook the form dropdowns up so they show real data from the API
* [x] Borrow the validation rules from wizard steps
* [x] Make the form submission submit data to the CRM API
* [x] Deal with the API response appropriately
  * [x] Show success if a matching record **doesn't exists**
  * [x] Show an extra verification page if a matching record **does exist**
* [x] Improve the routes, should be `mailing-list/signup` or similar
* [x] Add specs, maybe (not sure at the commitment level yet)
  * [x] Test the model and validation
  * [x] Test the controller and calling the API

### Final bits of tidying up

* [x] Link to privacy policy from the checkbox label
* [x] More specs covering methods in `MailingList::Signup`
* [x] Move postcode down to the bottom of the form
* [x] Ensure relevant values are present in unverified submission
* [x] Ensure the `candidate_id` and `qualification_id` are present in verified submission

## Progress

| Form | Form with [errors](https://github.com/DFE-Digital/get-into-teaching-app/pull/1165/files#diff-44438ce218f5287c58d0017f965d888715635d94280669896f75841fbd7b4cd7R184) |
| ------ | ------- |
| Screenshot from 2021-04-26 16-46-12  | Screenshot from 2021-04-26 16-47-43 |

### CODE REVIEW

File: app/models/mailing_list/signup.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
+require "attribute_filter"
+
+module MailingList
+  class Signup
+    include ActiveModel::Model
+    include ActiveModel::Validations::Callbacks
+    include ::Wizard::ApiClientSupport
```
Comment (ethax-ross) [inline app/models/mailing_list/signup.rb:7]: Can be one for later, but I think we should rename this to something more generic like `Form::ApiClientSupport`

File: app/models/mailing_list/signup.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    end
+
+    def strip_whitespace
+      email&.strip!
+      first_name&.strip!
+      last_name&.strip!
+    end
+
+    def convert_string_ids_to_integers
+      self.degree_status_id = degree_status_id.to_i
```
Comment (ethax-ross) [inline app/models/mailing_list/signup.rb:133]: If we uses `attribute` instead of `attr_accessor` we wouldn't need these as we can type it as `:integer`?

**Review by ethax-ross — DISMISSED** [review]
Looks great -- couple of take it or leave it comments; I would perhaps exclude the route from prod before this is shipped though/until the remaining implementation is done? I know its not going to be indexed but you never know... 🙄

File: app/models/mailing_list/signup.rb (hunk 2 above)
Comment (peteryates) [reply app/models/mailing_list/signup.rb:133]: Ah good call, will do.

**Comment by MylesJarvis** [comment]
@peteryates I think we should perhaps keep the order the same as the current mailing list sign up, which leaves postcode to the end. Thoughts?

File: app/controllers/mailing_list_signups_controller.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    @signup = MailingList::Signup.new
+  end
+
+  def create
+    @signup = MailingList::Signup.new(new_signup_params)
+
+    if @signup.valid?
+      save_signup_to_session
+
+      return redirect_to(edit_mailing_list_signup_path) if @signup.already_signed_up?
```
Comment (ethax-ross) [inline app/controllers/mailing_list_signups_controller.rb:17]: This is a bit misleading I think - `already_signed_up?` will return `true` if the candidate already exists, not if they exist _and_ they have already signed up to the mailing list

File: app/controllers/mailing_list_signups_controller.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    session.delete(SESSION_STORE_KEY)
+    redirect_to mailing_list_step_path(path)
+  end
+
+  def restore_signup_from_session
+    signup_data = session[SESSION_STORE_KEY]
+
+    if signup_data.nil?
+      Rails.logger.warn("tried to confirm identity with blank session")
+      return redirect_to(new_mailing_list_signup_path)
```
Comment (ethax-ross) [inline app/controllers/mailing_list_signups_controller.rb:59]: Could this not result in odd bugs if we return a `redirect_to` the caller will continue to process the code after the call to `restore_signup_from_session` (so line 33 onward in this file)?

File: app/models/mailing_list/signup.rb [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+    validate :timed_one_time_password_is_correct,
+             on: :verify
+
+    def degree_status_options
+      @degree_status_options ||= query_degree_status
+    end
+
+    def teaching_subjects
+      @teaching_subjects ||= [OpenStruct.new(id: nil, value: "Please select")] + query_teaching_subjects
```
Comment (ethax-ross) [inline app/models/mailing_list/signup.rb:78]: Not a big deal, but could avoid having to do this by using `prompt: { "Please select" }` in the view I think

File: app/models/mailing_list/signup.rb [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+      GetIntoTeachingApiClient::MailingListApi.new
+    end
+
+    def timed_one_time_password_is_correct
+      request = GetIntoTeachingApiClient::ExistingCandidateRequest.new(identity_data)
+      response = api.exchange_access_token_for_mailing_list_add_member(timed_one_time_password, request)
+
+      Rails.logger.info("MailingList::Wizard#exchange_access_token: #{AttributeFilter.filtered_json(response)}")
+
+      response.tap do |r|
```
Comment (ethax-ross) [inline app/models/mailing_list/signup.rb:186]: I don't feel strongly about it, but this method is doing quite a lot and more than the method name suggests. I wonder if it would be clearer to split this and have an `after_validation :prepopulate` that sets the necessary ids etc from the totp response? You then have to track the response as an instance variable, so maybe what you have is more pragmatic 🤷‍♂️

File: spec/models/mailing_list/signup_spec.rb [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    describe "#degree_status_id" do
+      let(:degree_statuses) { GetIntoTeachingApiClient::PickListItemsApi.new.get_qualification_degree_status }
+      let(:message) { "Select your degree stage from the list" }
+
+      it { is_expected.to validate_presence_of(:degree_status_id).with_message(message) }
+
+      describe "values" do
+        specify "allows teaching subject values" do
+          degree_statuses.map(&:id).each do |ds|
+            expect(subject).to allow_value(ds).for(:degree_status_id)
```
Comment (ethax-ross) [inline spec/models/mailing_list/signup_spec.rb:97]: You can pass multiple values to these shoulda matchers with `allow_values` to make them more concise:

`it { is_expected.to allow_values(degree_statuses.map(&:id)).for(:degree_status_id) }`

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -8,6 +8,7 @@
 end
 
 require "rspec/rails"
+require "active_record"
```
Comment (ethax-ross) [inline spec/rails_helper.rb:11]: Why was this necessary?

**Review by ethax-ross — COMMENTED** [review]
Looks good overall, just some minor questions/observations, mostly take it or leave it

File: app/controllers/mailing_list_signups_controller.rb (hunk 3 above)
Comment (peteryates) [reply app/controllers/mailing_list_signups_controller.rb:17]: Yeah - I named it when I didn't fully understand what was going on. Renamed to `#exists_in_crm?`.

File: app/controllers/mailing_list_signups_controller.rb (hunk 4 above)
Comment (peteryates) [reply app/controllers/mailing_list_signups_controller.rb:59]: Yep agree 🤦🏽 

Reworked with a VerifyWithEmptySession error

File: app/controllers/mailing_list_signups_controller.rb (hunk 3 above)
Comment (ethax-ross) [reply app/controllers/mailing_list_signups_controller.rb:17]: That kind of reads like the same thing when you call it on `@signup.exists_in_crm?` 😬 maybe something like `@signup.existing_candidate?`

File: spec/rails_helper.rb (hunk 8 above)
Comment (peteryates) [reply spec/rails_helper.rb:11]: When switching from `attr_accessor` to `attribute` I'm getting errors from shoulda (originating [here](https://github.com/thoughtbot/shoulda-matchers/blob/master/lib/shoulda/matchers/rails_shim.rb#L51)) matchers unless ActiveRecord is present.

Comment (ethax-ross) [reply spec/rails_helper.rb:11]: I think you're maybe just missing an `include ActiveModel::Attributes` on your signup class?

Comment (ethax-ross) [reply spec/rails_helper.rb:11]: Ah fair enough, weird the step models work but this one doesn't 🤔 I can't see any differences between them though 🤷

File: app/models/mailing_list/signup.rb (hunk 6 above)
Comment (peteryates) [reply app/models/mailing_list/signup.rb:186]: Yeah I did try with a separate method but the logic became a bit hard to follow. I'll rename the method to add a bit more clarity

File: spec/rails_helper.rb (hunk 8 above)
Comment (peteryates) [reply spec/rails_helper.rb:11]: Nor me, I spent ages comparing them side by side 🤷🏽‍♂️

**Review by ethax-ross — APPROVED** [review]
Looks good

**Review by ethax-ross — APPROVED** [review]
